docs: add HIP-1195 Hiero Hooks docs (#362)
hashgraph/hedera-docs: hedera/core-concepts/accounts/hiero-hooks.mdx

| Field | Description |
|------|-------------|
| **Hook ID** | Identifies the EVM Hook whose storage is being updated, including the owning entity (account or contract) and the hook’s 64-bit ID. |
| **Storage Updates** | A list of updates applied to the hook’s persistent storage. Supports direct slot updates (`EvmHookStorageSlot`) and mapping updates (`EvmHookMappingEntries`). |

---

### Hook Storage Details

Hooks can maintain state in their own storage. The `HookStoreTransaction` allows for granular updates to this storage without executing the hook's full logic.
 * **Slots**: Write or delete a 32-byte key → 32-byte value via EvmHookStorageSlot.
 * **Mappings**: Update entries under a mapping slot via EvmHookMappingEntries, either by explicit key or by providing the preimage whose Keccak256 hash yields the key.

<Note>
    **Important constraints**

    When deleting a hook, you must first clear all its storage slots. Otherwise, the deletion will fail with the status `HOOK_DELETION_REQUIRES_ZERO_STORAGE_SLOTS`. Additionally, an account cannot be deleted if it has any non-zero hooks attached. `CryptoDelete` transaction fails with `TRANSACTION_REQUIRES_ZERO_HOOKS`.
</Note>

---

> ### Extension Points

Reviewer: We might want to add a line that Users can propose a new Extension Point through the Hiero HIP process.

### Step 2: Attach the Hook

Attach the hook using:

- `AccountCreateTransaction` or `ContractCreateTransaction` for new entities  
- `AccountUpdateTransaction` or `ContractUpdateTransaction` for existing entities  

You must specify:

- Extension point  
- Hook ID (64-bit identifier)  
- EVM Hook (the deployed `ContractId`)  

### Step 3: Trigger the Hook

**Hooks execute ONLY when explicitly referenced in the `TransferTransaction`**—attachment alone does not trigger execution.

When triggered:

1. Transaction execution pauses  
2. Hook logic runs  
3. Transaction continues only if the hook returns `true`  

If the hook reverts or returns `false`, the entire transaction fails and all state changes roll back.

---

## Account Allowance Hooks

Account Allowance Hooks validate allowance-based transfers during a `TransferTransaction`. **They are only invoked if the hook is explicitly referenced/specified in the transaction for the allowance.**

They follow the same execution and rollback guarantees described in the lifecycle.

---

### Core Interface

```solidity
interface IHieroAccountAllowanceHook {

>     function allow(

Reviewer 2: This interface has 3 methods I believe and not just `allow`:
- allowPre(HookContext, ProposedTransfers) - called before the transfer
- allowPost(HookContext, ProposedTransfers) - called after the transfer                                             
- allow(HookContext, ProposedTransfers) - inherited from base IHieroHook  

It looks like further below, we reference `allowPre` and `allowPost` but I think we need to define it here as well?

3. NFT transfer's `allowPre(...)`.
4. Fungible token transfer's `allowPre(...)`.
5. (Main transaction logic)
6. NFT transfer's `allowPost(...)`.
7. Fungible token transfer's `allowPost(...)`.

### Limits

- **Hook invocations per transaction**: Each transaction can invoke a maximum of 10 hooks (`hooks.maxHookInvocationsPerTransaction=10`)
- **Child records**: Child records generated by hook calls are capped at 50 (`consensus.handle.maxFollowingRecords=50`)

---

## Hooks vs Smart Contracts

| Feature           | Account Hooks                           | Smart Contracts            |
| ----------------- | ------------------------------------- | -------------------------- |
| Primary Use       | Inline validation for native services | Full on-chain applications |
| Trigger           | **By reference** in a transaction (e.g., `CryptoTransfer`) | Explicit `ContractCall`    |
| Execution Context | `0x16d` with owner privileges         | Contract address           |
| State Updates     | `HookStoreTransaction`             | Full contract execution    |
| Cost              | Lower                                 | Higher                     |

---

## Next Steps

> * [HookStoreTransaction](#) - SDK documentation

Reviewer 2: Do we need to put an actual link here?

Unlike regular smart contracts, hooks execute in a special EVM context where `address(this)` is always the reserved system address `0x16d`, enabling them to act with the privileges of the account they're attached to. This model combines smart contract flexibility with native HAPI transaction efficiency, allowing custom validation without deploying full-scale contracts.

## Core Concepts

Hiero Hooks are a mechanism for [**Account Abstraction**](/hedera/support-and-community/glossary#account-abstraction) on Hedera, enabling custom validation and logic without migrating entire applications to the EVM. A hook is a small piece of Solidity logic that is **triggered only when referenced/specified in a `TransferTransaction`**—not automatically.

Think of it like a webhook for the ledger itself. Instead of waiting for an off-chain call, the hook runs inside the network when a transaction explicitly references it. Hooks can check conditions before execution, update state, log data, or stop a transfer if validation fails.

> ### Why Account Hooks?

Reviewer 3: I would lean towards just saying `Hooks` instead of Account Hooks. The first hook we implemented is `Account Allowance` hook, but in the future there will be many other types of hooks.

> Before Account Hooks, developers faced two major limitations:

Reviewer 3: ```suggestion
Before Hooks, developers faced two major limitations:
```

1. **Protocol dependency**: New functionality required network-wide upgrades through HIPs (slow and heavyweight)
2. **EVM migration**: Moving applications to smart contracts sacrificed the performance and cost-efficiency of native HAPI transactions

> Account Hooks solve this by allowing developers to inject custom logic directly into native flows, offering better performance and lower cost than general-purpose `ContractCall` operations.

Reviewer 3: ```suggestion
Hooks solve this by allowing developers to inject custom logic directly into native flows, offering better performance and lower cost than general-purpose `ContractCall` operations.
```

### Key Characteristics

| Concept | Description |
|:--------|:------------|
| **Trigger Model** | **Triggered only when referenced/specified** in a `TransferTransaction`—not automatic event listeners. |
| **Implementation** | EVM Hooks: Solidity contracts executed by the network's EVM. |
| **Extension Point** | Account Allowance Hooks validate token transfers that consume an allowance. |
| **Key Advantage** | Custom logic on native assets (HBAR and HTS tokens) without `ContractCall` overhead. |
| **Use Cases** | Compliance rules, transfer constraints, custom validation logic. |
---

## How Account Hooks Work

> An Account Hook is a small piece of EVM bytecode, implemented as a Solidity contract, that is attached to a Hedera entity such as an account or contract.

Reviewer 3: ```suggestion
A Hook is a small piece of EVM bytecode, implemented as a Solidity contract, that is attached to a Hedera entity such as an account or contract.
```

- **Privileges**  
  The hook executes with the owner’s privileges. A hook attached to account `0.0.123` can act as `0.0.123`.
- **Identity**  
  - `address(this)` → `0x16d`  
  - `msg.sender` → transaction payer  
- **Storage**
  The storage used during hook execution is "Hook" storage, instead of the contract's storage. This ensures that the hook's state is isolated and managed specifically for the hook instance.

This is Hedera’s implementation of account abstraction.

---

### Hook Management and State

Hooks are managed via standard HAPI transactions and can maintain their own persistent state.

#### Hook Management Transactions

| Transaction Type | Purpose | Key SDK Class |
|------------------|---------|---------------|
| **Creation** | Attach a hook to a new or existing entity (account or contract). | `CryptoCreateTransaction`, `ContractCreateTransaction` |
| **Update / Deletion** | Modify hook parameters or remove the hook from an entity. | `AccountUpdateTransaction`, `ContractUpdateTransaction` |
| **Storage Update** | Update a hook’s persistent storage without executing its logic. | `HookStoreTransaction` |

---

### HookStoreTransaction

> The `HookStoreTransaction` enables fast, low-overhead updates to a hook’s storage. This allows hook configuration changes such as passcodes or allowlists without the cost of a full `ContractCall`.

Reviewer 3: ```suggestion
The `HookStoreTransactionBody` enables fast, low-overhead updates to a hook’s storage. This allows hook configuration changes such as passcodes or allowlists without the cost of a full `ContractCall`.
```

#### HookStoreTransaction Properties

| Field | Description |
|------|-------------|
| **Hook ID** | Identifies the EVM Hook whose storage is being updated, including the owning entity (account or contract) and the hook’s 64-bit ID. |

> | **Storage Updates** | A list of updates applied to the hook’s persistent storage. Supports direct slot updates (`LambdaStorageSlot`) and mapping updates (`LambdaMappingEntries`). |

Reviewer 3: ```suggestion
| **Storage Updates** | A list of updates applied to the hook’s persistent storage. Supports direct slot updates (`EvmHookStorageSlot`) and mapping updates (`EvmHookMappingEntries`). |
```

 * **Mappings**: Update entries under a mapping slot via EvmHookMappingEntries, either by explicit key or by providing the preimage whose Keccak256 hash yields the key.

<Note>
    **Important constraints**

    When deleting a hook, you must first clear all its storage slots. Otherwise, the deletion will fail with the status `HOOK_DELETION_REQUIRES_ZERO_STORAGE_SLOTS`. Additionally, an account cannot be deleted if it has any non-zero hooks attached. `CryptoDelete` transaction fails with `TRANSACTION_REQUIRES_ZERO_HOOKS`.
</Note>

---

### Extension Points

Hooks attach to specific extension points in a transaction's lifecycle. An extension point defines the type of hook allowed for a transaction but doesn't specify when or why a hook is activated.

Currently, the first supported extension point is the Account Allowance Hook (`ACCOUNT_ALLOWANCE_HOOK`). This hook runs when a `TransferTransaction` attempts to spend an allowance from an account.

Future extension points may include other native transaction types or entity lifecycle events, enabling hooks to validate or augment a wide range of on-chain operations.

---

## **Hook Lifecycle**

> Understanding how hooks are deployed, attached, and executed is essential for using Account Hooks effectively.

Reviewer 3: ```suggestion
Understanding how hooks are deployed, attached, and executed is essential for using Hooks effectively.
```

The execution phases are as follows:

1. **Pre-Hooks Execution**
   
   a. All hooks of type `PRE_HOOK` are executed. The order is determined by the sequence of transfers (e.g., NFT transfers vs. fungible token transfers) in the `TokenTransfers` list.

2. **Pre-Post Hooks (Pre-Transfer Part)**
   
   a. The `allowPre(...)` function of all pre-post hooks is executed. Again, the execution order follows the `TokenTransfers` list.

3. **Pre-Post Hooks (Post-Transfer Part)**
   
   a. After the main transaction logic has completed, the `allowPost(...)` function of all pre-post hooks is executed, following the `TokenTransfers` list order.

For example, if a transaction's `TokenTransfers` list specifies an NFT transfer before a fungible token transfer, and both have `PRE_HOOK` and pre-post hooks, the execution would be:

1. NFT transfer's `PRE_HOOK`.
2. Fungible token transfer's `PRE_HOOK`.
3. NFT transfer's `allowPre(...)`.
4. Fungible token transfer's `allowPre(...)`.
5. (Main transaction logic)
6. NFT transfer's `allowPost(...)`.
7. Fungible token transfer's `allowPost(...)`.

### Limits

> 

Reviewer 3: We can also add, Each transaction can only allow 10 hook invocations. That is `hooks.maxHookInvocationsPerTransaction` config

- **Hook invocations per transaction**: Each transaction can invoke a maximum of 10 hooks (`hooks.maxHookInvocationsPerTransaction=10`)
- **Child records**: Child records generated by hook calls are capped at 50 (`consensus.handle.maxFollowingRecords=50`)

---

## Hooks vs Smart Contracts

> | Feature           | Account Hooks                           | Smart Contracts            |

Reviewer 3: ```suggestion
| Feature           | Hooks                           | Smart Contracts            |
```
